src/adopt.rs

```diff
@@ -78,6 +87,47 @@ pub unsafe trait Adopt: sealed::Sealed {
 /// Implementation of [`Adopt`] for [`Rc`] which enables `Rc`s to form a cycle
 /// of strong references that are reaped by `Rc`'s [`Drop`] implementation.
 unsafe impl<T> Adopt for Rc<T> {
+    /// `Rc`'s inner owned value. For an `Rc<T>`, the inner owned value is a
+    /// `T`.
+    type Inner = T;
+
+    /// TODO: document me!
+    fn adopt(this: &mut Self, other: &Self)
```

Comment:
maybe adopt should return `Result<(), &Self>` where the error variant contains `other` if the adoption failed?

Reply:
Or maybe there is `try_adopt` and `adopt` which unwraps `try_adopt`.

Reply:
1. The `E` type in `Result` should implement `Error`, otherwise it won't compose well with existing error-handling mechanisms, such as the `?` sugar.
2. `try_adopt` that returns a `Result` and `adopt` that unwraps the former is indeed a good, ergonomic API. The standard library does that a lot too, I think.
3. Why would you want to return `other` in case of a failure, if it's a reference and the callee already has it anyway?